src/lib/OccAnalysis.hs

```diff
@@ -0,0 +1,559 @@
+      -- correspond to the free vars of the Abs of the block answer,
+      -- by the decls traversed so far. occNest takes care to uphold
+      -- this invariant, but we temporarily reset the state to an
+      -- empty map, just so that names from the surrounding block
+      -- don't end up influencing elimination decisions here. Note
+      -- that we restore the state (and accumulate free vars of the
+      -- analyzed block into it) right after occNest.
+      Abs decls' ans' <- locally $ occNest a decls ans
+      ty' <- occ accessOnce ty  -- TODO Is free variable counting enough here?
+      countFreeVarsAsOccurrences effs
+      return $ Block (BlockAnn ty' effs) decls' ans'
+
+-- TODO(optimization) Could reuse the free variable caching from dce here too.
+
+data ElimResult n where
+  ElimSuccess :: Abs (Nest Decl) Atom n -> ElimResult n
+  ElimFailure :: Decl n l -> UsageInfo -> Abs (Nest Decl) Atom l -> ElimResult n
+
+occNest :: Access n -> Nest Decl n l -> Atom l
+        -> OCCM n (Abs (Nest Decl) Atom n)
+occNest a decls ans = case decls of
+  Empty -> Abs Empty <$> occ a ans
+  Nest d@(Let _ binding) ds -> do
+    isPureDecl <- isPure binding
+    -- Note that we only ever analyze the abs below under this refreshAbs,
+    -- which will remove any references to b upon exit (it happens
+    -- because refreshAbs of StateT1 triggers hoistState, which we
+    -- implement by deleting the entries that can't hoist).
+    dceAttempt <- refreshAbs (Abs d (Abs ds ans))
+      \d'@(Let b' (DeclBinding _ _ expr')) (Abs ds' ans') -> do
+        exprIx <- summaryExpr $ sink expr'
+        extend b' exprIx do
+          below <- locally $ occNest (sink a) ds' ans'
+          accessInfo <- getAccessInfo $ binderName d'
+          let usage = usageInfo accessInfo
+          let dceAttempt = case isPureDecl of
+               False -> ElimFailure d' usage below
+               True  ->
+                 -- Or hoistUsingCachedFVs in the monad, if we decide to do
+                 -- that optimization
+                 case hoist d' below of
+                   HoistSuccess below' -> ElimSuccess below'
+                   HoistFailure _ -> ElimFailure d' usage below
+          return dceAttempt
+    case dceAttempt of
+      ElimSuccess below' -> return below'
+      ElimFailure (Let b' binding') usage (Abs ds'' ans'') -> do
+        -- Using accessOnce here, instead of the computed Access for
+        -- the decl's binder.  This means that variable bindings cut
+        -- occurrence analysis, and each binding is considered for
+        -- inlining separately.
+        DeclBinding _ ty expr <- occ accessOnce binding'
+        let binding'' = DeclBinding (OccInfo usage) ty expr
+        return $ Abs (Nest (Let b' binding'') ds'') ans''
+
+instance HasOCC DeclBinding where
+  occ a (DeclBinding ann ty expr) = do
+    expr' <- occ a expr
+    -- TODO(optimization) Is free variable counting enough for types?
```

Comment:
How is that related to optimization? Seems more like a correctness concern? And I think the answer is as above.

Reply:
I tagged it "optimization" because presumably traversing something to count its free variables is easier than running occurrence analysis on it.  May also affect the precision; but I want to learn more about the right treatment of types.

Reply:
Subsumed by the new `occTy` helper.